accessible/tree.service.js

```diff
@@ -0,0 +1,325 @@
+            // if the previous key is shift, we're shift-tabbing mode
+            this.goToPreviousTree(treeId, e);
+          } else {
+            // if previous key isn't shift, we're tabbing
+            this.goToNextTree(treeId, e);
+          }
+          break;
+      }
+      this.previousKey_ = e.keyCode;
+    },
+    goToNextTree: function(treeId, e) {
+      var next = false;
+      for (var i = 0; i < this.trees.length; i++) {
+        if (next) {
+          this.trees[i].focus();
+          e.preventDefault();
+          e.stopPropagation();
+          return;
+        }
+        if (this.trees[i].id == treeId) {
+          next = true;
+        }
+      }
+      e.preventDefault();
+      e.stopPropagation();
+    },
+    goToPreviousTree: function(treeId, e) {
+      if (treeId == this.trees[0].id) {
+        return;
+      }
+      // otherwise, go to the previous tree class
+      var next = false;
+      for (var i = (this.trees.length - 1); i >= 0; i--) {
+        if (next) {
+          this.trees[i].focus();
+          e.preventDefault();
+          e.stopPropagation();
+          return;
+        }
+        if (this.trees[i].id == treeId) {
+          next = true;
+        }
+      }
+      e.preventDefault();
+      e.stopPropagation();
+    },
+    keyHandler: function(e, tree) {
+      var treeId = tree.id;
+      var node = this.getActiveDesc(treeId);
+      if (!node) {
+        blocklyApp.debug && console.log('KeyHandler: no active descendant');
+      }
+      blocklyApp.debug && console.log(e.keyCode);
+      blocklyApp.debug && console.log('inside TreeService');
+      switch (e.keyCode) {
+        case 9:
+          // 16,9: shift, tab
+          if (e.shiftKey) {
+            console.log('shifttabbing');
+            // if the previous key is shift, we're shift-tabbing mode
+            this.goToPreviousTree(treeId, e);
+          } else {
+            // if previous key isn't shift, we're tabbing
+            // we want to go to the run code button
+            this.goToNextTree(treeId, e);
+          }
+          break;
+        case 37:
+          // left-facing arrow: go out a level, if possible. If not, do nothing
+          e.preventDefault();
+          e.stopPropagation();
+          blocklyApp.debug && console.log('in left arrow section');
+          var nextNode = node.parentNode;
+          if (node.tagName == 'BUTTON' || node.tagName == 'INPUT') {
+            nextNode = nextNode.parentNode;
+          }
+          while (nextNode && nextNode.className != 'treeview' &&
+                nextNode.tagName != 'LI') {
+            nextNode = nextNode.parentNode;
+          }
+          if (nextNode.className == 'treeview' || nextNode == null) {
+            return;
+          }
+          this.updateSelectedNode(nextNode, tree);
+          break;
+        case 38:
+          // up-facing arrow: go up a level, if possible. If not, do nothing
+          e.preventDefault();
```

Comment:
Incidentally, it would be nice to have a convention for whether these are placed at the beginning or at the end. Here, it's the former; in goToNextTree(), it's the latter. Can we standardize?